### PR TITLE
Copy number segment data API and its unit test

### PR DESCRIPTION
**Pull request by JiaoJiao123**
# What? Why?
The basic usage of this API is to fetch copy number segment data and the details is specified in this [RFC ](https://docs.google.com/document/d/1mqqnyDeaWN0uTDuBU2bzUBFwetUWHX_PtHstCeCpnLk/edit#heading=h.pm7gdme3529l)
It is designed for [PR#1679 implementing IGV tab with its javascript version](https://github.com/cBioPortal/cbioportal/pull/1679). 

Changes proposed in this pull request:
- query copy number segment data by cancer study, sample list and hugo symbol
- related unit test

# Notify reviewers
@sheridancbio 
@cBioPortal/backend

### CODE REVIEW

**Comment by ecerami** [comment]
Do these API change mesh with the recent API work that @ersinciftci, @sheridancbio and others have been working on?  I just want to make sure that we are only pushing out API additions that mesh with the larger API strategy.

**Comment by JiaoJiao123** [comment]
@ecerami It should be. I discussed with @sheridancbio about the API structure to use and etc. I think we are using sprint 2/3.

**Comment by sheridancbio** [comment]
@ecerami @ersinciftci We are separating the api controllers into two packages .. org.cbioportal.web will contain controllers which have the REST design approach which @ersinciftci is leading the design on. Older controllers and controllers which are written initially in a non-REST idiom will go in org.cbioportal.weblegacy. This endpoint that @JiaoJiao123 has developed is needed now, and it is being deployed in the org.cbioportal.weblegacy package. There are two other data services that also being deployed at this time: https://github.com/cBioPortal/cbioportal/pull/1618 and https://github.com/cBioPortal/cbioportal/pull/1648/files. When the new API design effort has stabilized, we can convert these data services into the REST based interaction. The REST Api design effort is currently focussed on the patient-view functionality. It does have an api endpoint for genomic segment data defined. Possibly when it comes time to develop the backend implementation for the new REST endpoint we can simply convert the persistence layer and service layer code in this PR to provide the data services for the new endpoint.

**Comment by pieterlukasse** [comment]
maybe good to add an integration test as well (once #1513 is merged...almost there)

File: persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/CNASegmentMapper.xml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+<!DOCTYPE mapper PUBLIC "-//mybatis.org//DTD Mapper 3.0//EN" "http://mybatis.org/dtd/mybatis-3-mapper.dtd">
+
+<mapper namespace="org.cbioportal.persistence.mybatis.CNASegmentMapper">
+    <cache/>
+
+    <select id="getCNASegmentData" resultType="org.cbioportal.model.CNASegmentData">
+        select s.stable_id as sample, cns.chr as chr, cns.start as start, cns.end as end, cns.num_probes as numProbes, cns.segment_mean as value from sample s, copy_number_seg cns, cancer_study cs 
+        where cs.cancer_study_identifier = #{cancerStudyId} 
+        
+        <if test="sampleIds != null and !sampleIds.isEmpty()">
```
Comment (sheridancbio) [inline persistence/persistence-mybatis/src/main/resources/org/cbioportal/persistence/mybatis/CNASegmentMapper.xml]: Do we really want to allow a user to get the *complete* set of copy number segments for a study?  I guess it depends on how big this return of data could be. If it is many megabytes for a big study, perhaps we want to require a gene list or a sample list?

File: service/src/main/java/org/cbioportal/service/impl/CNASegmentServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Autowired
+    private CNASegmentRepository cnaSegmentRepository;
+    
+    @Autowired
+    private GeneRepository geneRepository;
+
+    @Override
+    public List<CNASegmentData> getCNASegmentData(String cancerStudyId, List<String> hugoGeneSymbols, List<String> sampleIds) {
+        List<String> chromosomes = new ArrayList<String>();
```
Comment (sheridancbio) [inline service/src/main/java/org/cbioportal/service/impl/CNASegmentServiceImpl.java]: wouldn't we expect duplicates here?  Shouldn't those duplicates be merged into a unique list?  I think Set<String> chromosomes = new Set<String>()  might be what we want.

File: web/src/main/java/org/cbioportal/web/config/CustomObjectMapper.java [hunk 3]
```diff
@@ -28,6 +26,5 @@ public CustomObjectMapper() {
         mixinMap.put(Sample.class, SampleMixin.class);
         mixinMap.put(TypeOfCancer.class, TypeOfCancerMixin.class);
         super.setMixInAnnotations(mixinMap);
-        super.configure(JsonGenerator.Feature.WRITE_NUMBERS_AS_STRINGS, true);
```
Comment (sheridancbio) [inline web/src/main/java/org/cbioportal/web/config/CustomObjectMapper.java]: we must test the built warfile and make sure all visualizations work after making this change.

File: web/src/main/java/org/cbioportal/weblegacy/CNASegmentController.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.List;
+import org.cbioportal.model.CNASegmentData;
+import org.cbioportal.service.CNASegmentService;
+
+@RestController
+public class CNASegmentController {
+
+    @Autowired
+    private CNASegmentService cnaSegmentService;
+
```
Comment (sheridancbio) [inline web/src/main/java/org/cbioportal/weblegacy/CNASegmentController.java]: Add some human readable descriptions with @ApiOperation and @ApiParam annotations -- see GenePanelController.java for examples

File: web/src/test/java/org/cbioportal/web/api/ApiControllerTest.java [hunk 5]
```diff
@@ -32,11 +32,11 @@
 import org.springframework.test.context.web.WebAppConfiguration;
 import org.springframework.test.web.servlet.MockMvc;
 import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
-import org.springframework.test.web.servlet.result.MockMvcResultHandlers;
 import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
 import org.springframework.test.web.servlet.setup.MockMvcBuilders;
 import org.springframework.web.context.WebApplicationContext;
 
+@Ignore
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/web/api/ApiControllerTest.java]: I had the impression that this went by the test methods .. not the entire controller test class

File: web/src/test/java/org/cbioportal/weblegacy/CNASegmentControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        cNASegmentData1.setSample("TCGA-AG-3732-01");
+        
+        CNASegmentData cNASegmentData2 = new CNASegmentData(); 
+        cNASegmentData2.setStart(148347132);
+        cNASegmentData2.setEnd(148348416);
+        cNASegmentData2.setChr("7");
+        cNASegmentData2.setNumProbes(3);
+        cNASegmentData2.setValue((float) -1.5009);
+        cNASegmentData2.setSample("TCGA-AG-3732-01");
+
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/weblegacy/CNASegmentControllerTest.java]: two blank lines here is excessive --- consider having no blank line here or between other statements within this function.

File: web/src/test/java/org/cbioportal/weblegacy/CNASegmentControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        CNASegmentData cNASegmentData2 = new CNASegmentData(); 
+        cNASegmentData2.setStart(148347132);
+        cNASegmentData2.setEnd(148348416);
+        cNASegmentData2.setChr("7");
+        cNASegmentData2.setNumProbes(3);
+        cNASegmentData2.setValue((float) -1.5009);
+        cNASegmentData2.setSample("TCGA-AG-3732-01");
+       
+        
+      	mockResponse.add(cNASegmentData1);
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/weblegacy/CNASegmentControllerTest.java]: this line should be indented the same as others

**Comment by JiaoJiao123** [comment]
@pieterlukasse that's a good suggestion. But I might add the integration test later.

File: web/src/main/java/org/cbioportal/weblegacy/GeneController.java [hunk 8]
```diff
@@ -49,8 +52,13 @@
     @Autowired
     private GeneService geneService;
 
+    @ApiOperation(value = "Get all information about genes from hugo symbols including entrezGeneId, hugoGeneSymbol, cytoband, length and chromosome",
+            nickname = "getGeneListByHugoSymbols",
+            notes = "")
+    @Transactional
     @RequestMapping(method = {RequestMethod.GET, RequestMethod.POST}, value = "/gene/fetch-by-hugo")
-    public List<Gene> getGeneListByHugoSymbols(@RequestParam(required = true) List<String> hugo_gene_symbols) {
+    public List<Gene> getGeneListByHugoSymbols(@ApiParam(value = "a list of gene hugo symbols that you want to query on")
```
Comment (sheridancbio) [inline web/src/main/java/org/cbioportal/weblegacy/GeneController.java]: An example case .. or directions to the /api/gene endpoint would help the user understand better

File: web/src/test/java/org/cbioportal/weblegacy/CNSegmentControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        cnSegmentData1.setSample("TCGA-AG-3732-01");
+        CNSegmentData cnSegmentData2 = new CNSegmentData(); 
+        cnSegmentData2.setStart(148347132);
+        cnSegmentData2.setEnd(148348416);
+        cnSegmentData2.setChr("7");
+        cnSegmentData2.setNumProbes(3);
+        cnSegmentData2.setValue((float) -1.5009);
+        cnSegmentData2.setSample("TCGA-AG-3732-01");
+       
+      	mockResponse.add(cnSegmentData1);
```
Comment (sheridancbio) [inline web/src/test/java/org/cbioportal/weblegacy/CNSegmentControllerTest.java]: indent

**Pull request merged**